narwhal/executor/src/subscriber.rs

```diff
@@ -243,148 +274,222 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
-            let batch_fetch_duration = payload.metadata.created_at.elapsed().as_secs_f64();
-            self.metrics
-                .batch_execution_latency
-                .observe(batch_fetch_duration);
-            debug!(
-                "Batch {:?} took {} seconds to be fetched for execution since creation",
-                payload.digest(),
-                batch_fetch_duration
-            );
-            return payload;
-        }
-        let _timer = self.metrics.subscriber_remote_fetch_latency.start_timer();
-        let mut stagger = Duration::from_secs(0);
-        let mut futures = vec![];
-        for worker in workers {
-            let future = self.fetch_from_worker(stagger, worker, digest);
-            futures.push(future.boxed());
-            // TODO: Make this a parameter, and also record workers / authorities that are down
-            //       to request from them batches later.
-            stagger += Duration::from_millis(200);
+    ) -> HashMap<BatchDigest, Batch> {
+        let mut fetched_batches = HashMap::new();
+
+        for (worker_id, digests) in batch_digests {
+            // TODO: Can further parallelize this by worker_id if necessary.
+            // Only have one worker for now so will leave this for a future
+            // optimization.
+            let mut remaining_digests = digests.clone();
+
+            let local_batches = self
+                .try_fetch_locally(remaining_digests.clone().into_iter().collect(), worker_id)
+                .await;
+            for (local_batch_digest, local_batch) in local_batches {
+                if remaining_digests.remove(&local_batch_digest) {
+                    let batch_fetch_duration =
+                        local_batch.metadata.created_at.elapsed().as_secs_f64();
+                    self.metrics
+                        .batch_execution_latency
+                        .observe(batch_fetch_duration);
+                    debug!(
+                        "Batch {:?} took {} seconds to be fetched for execution since creation",
+                        local_batch_digest, batch_fetch_duration
+                    );
+                    fetched_batches.insert(local_batch_digest, local_batch);
+                }
+            }
+
+            if remaining_digests.is_empty() {
+                return fetched_batches;
+            }
+
+            let _timer = self.metrics.subscriber_remote_fetch_latency.start_timer();
+            let mut stagger = Duration::from_secs(0);
+            let mut futures = vec![];
+            for worker in &workers {
+                let future = self.fetch_remote(stagger, worker.clone(), remaining_digests.clone());
+                futures.push(future.boxed());
+                // TODO: Make this a parameter, and also record workers / authorities that are down
+                // to request from them batches later.
+                stagger += Duration::from_millis(200);
```

Comment:
wondering if this stagger is low enough that we will be stepping on our own feet for large requests?

for example if it normally takes us (say) 1 sec to download 6MB, then by the end even if the first worker we contact would be saturating our link, we'd be clogging it up with 4 other requests for the same thing that waste a lot of resources, slow down the original request, and then get thrown out?

Reply:
I agree with your concern, but we may also wait too long in between worker fetches if we go to a bad node first. I think because we are more likely to hit a good worker than a bad worker, I will increase this to **500ms*** and we can track how that performs. 

*assuming 10mbps is a fair minimum network expectation